### PR TITLE
STAC-provided nodata value used as a fallback when the GeoTIFF doesn't provide one

### PR DESCRIPTION
**Pull request by alexgleith**
See bug in reported in `odc-stac` https://github.com/opendatacube/odc-stac/issues/259

The issue is that resolve_load_cfg sets `fill_value` from STAC metadata (nodata=1 for Landsat's qa_pixel), but leaves src_nodata_fallback=None. When _do_read calls `rasterio.warp.reproject`, it reads the source file's nodata tag (rdr.nodatavals → None for Landsat QA_PIXEL GeoTIFFs). With src_nodata=None, GDAL treats fill pixels as valid data, so they get warped into the output and block the fuser from using real data from the overlapping scene.

Edit to emphasise that this fix resolves a regression in behaviour between Rasterio/GDAL `1.5.0/3.12.2` and `1.4.3/3.9.x`

### CODE REVIEW

**Comment by Kirill888** [comment]
NO, fix your data files instead.

**Comment by alexgleith** [comment]
> NO, fix your data files instead.

There are some 10s of millions of data files hosted by the USGS. While I can report this, I can't get them fixed!

One file of the millions: `s3://usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF`

**Comment by Kirill888** [comment]
It takes good couple hours for a person with deep expertise to check for issues like that. It can take weeks of fluffing about by someone who is not as familiar with the domain. In early days it made sense for developers to debug these as they were possibly due to bad code. Now it's almost always bad data or bad python environment. So investment of developer time has zero reward in terms of code improvements for issues like this. Open source doesn't mean "free of charge support". And it certainly doesn't mean implementing dodgy workarounds for broken data sources that complicates code and leads to even more breakage in the future.

File: src/odc/loader/_reader.py [hunk 1]
```diff
@@ -170,6 +170,7 @@ def _resolve(
         return RasterLoadParams(
             _dtype(name, meta.data_type, "float32"),
             fill_value=_fill_value(meta),
+            src_nodata_fallback=_fill_value(meta),
```
Comment (Kirill888) [inline src/odc/loader/_reader.py]: NO, that doesn't do what you think. `_fill_value` will default to 0 for not configured non-float bands, but should remain as None if not set in STAC.

Comment (alexgleith) [reply src/odc/loader/_reader.py]: Would it be better to use `meta.nodata` here instead? Then it's `None` if it's not set in the STAC metadata.

Comment (Kirill888) [reply src/odc/loader/_reader.py]: I don't believe it should be enabled by default. It's a load time parameter that should be setable by the user at the time of calling `stac_load`, somehow... "For band XXX assume nodata value of YYY if it is not present on the data file for some silly reason". And remember that user CAN request output nodata to be different from nodata present in files/stac.

I don't know how your change interferes with the case when user DOES supply output nodata of some other value. You need way more tests than this for a MASSIVE behaviour change you are proposing. It fixes YOUR case, but what will it break?

Comment (Kirill888) [reply src/odc/loader/_reader.py]: Because next time, `nodata` won't be present in STAC either. I would say that STAC that DOES include `nodata` on an asset that has no `nodata` is an error. Why are we working around lack of data QA by a massive corporation that can clearly afford those checks?

Comment (Kirill888) [reply src/odc/loader/_reader.py]: Also is `nodata` actually absent in those TIFFs or is it simply "invisible by some versions of rasterio/gdal". In which case this work-around makes even less sense.

Comment (Kirill888) [reply src/odc/loader/_reader.py]: or even more likely "invisible in pypi based environment that includes wheels of various geo libs that ship incompatible versions of GDAL/proj libs that can not ever fit together into the same Python process without issues".

Comment (alexgleith) [reply src/odc/loader/_reader.py]: Good question about visibility. I used command line `rio` to check. I can check the other versions of Rio/GDAL.

Comment (Kirill888) [reply src/odc/loader/_reader.py]: yeah, that needs to happen before making a PR

**Comment by alexgleith** [comment]
For more information, I checked with two versions of GDAL. Neither show nodata values.

```bash
docker run --rm \
    --env AWS_REQUEST_PAYER=requester \
    --env AWS_ACCESS_KEY_ID=SECRET \
    --env AWS_SECRET_ACCESS_KEY=SECRET \
    ghcr.io/osgeo/gdal:ubuntu-small-3.12.2 \
    gdalinfo /vsis3/usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF

docker run --rm \
    --env AWS_REQUEST_PAYER=requester \
    --env AWS_ACCESS_KEY_ID=SECRET \
    --env AWS_SECRET_ACCESS_KEY=SECRET \
    ghcr.io/osgeo/gdal:ubuntu-small-3.9.3 \
    gdalinfo /vsis3/usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF
```

Both yield the same response:

```
Driver: GTiff/GeoTIFF
Files: /vsis3/usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF
Size is 7591, 7741
Coordinate System is:
PROJCRS["WGS 84 / UTM zone 48N",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["UTM zone 48N",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",105,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",500000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["Navigation and medium accuracy spatial referencing."],
        AREA["Between 102°E and 108°E, northern hemisphere between equator and 84°N, onshore and offshore. Cambodia. China. Indonesia. Laos. Malaysia - West Malaysia. Mongolia. Russian Federation. Singapore. Thailand. Vietnam."],
        BBOX[0,102,84,108]],
    ID["EPSG",32648]]
Data axis to CRS axis mapping: 1,2
Origin = (119385.000000000000000,275715.000000000000000)
Pixel Size = (30.000000000000000,-30.000000000000000)
Metadata:
  AREA_OR_POINT=Point
Image Structure Metadata:
  COMPRESSION=DEFLATE
  INTERLEAVE=BAND
Corner Coordinates:
Upper Left  (  119385.000,  275715.000) (101d34'42.02"E,  2d29'23.98"N)
Lower Left  (  119385.000,   43485.000) (101d34'53.30"E,  0d23'33.78"N)
Upper Right (  347115.000,  275715.000) (103d37'29.64"E,  2d29'37.47"N)
Lower Right (  347115.000,   43485.000) (103d37'34.18"E,  0d23'35.91"N)
Center      (  233250.000,  159600.000) (102d36' 9.84"E,  1d26'33.64"N)
Band 1 Block=256x256 Type=UInt16, ColorInterp=Gray
  Overviews: 3796x3871, 1898x1936, 949x968, 475x484, 238x242, 119x121
```

**Comment by alexgleith** [comment]
And for Rasterio:

```bash
python -c "import rasterio; print(rasterio.__version__); print(rasterio.__gdal_version__)" && AWS_REQUEST_PAYER=requester rio info s3://usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF | jq '.nodata'
```
> 1.4.3
> 3.9.3
> `null`

```bash
python -c "import rasterio; print(rasterio.__version__); print(rasterio.__gdal_version__)" && AWS_REQUEST_PAYER=requester rio info s3://usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_QA_PIXEL.TIF | jq '.nodata'
```

> 1.5.0
> 3.12.2
> `null`

**Comment by Kirill888** [comment]
Thanks for checking, I no longer have access to any kind of Sandbox on AWS to check data with requestor pays settings even if it's public.

This should be reported to data custodians of Landsat 9 on AWS.

As far as effective work-around within `odc-stac` are concerned it should not be "use value maybe set in STAC" it should be "use this value as a work-around for what I know is broken data". It's probably already settable by the user, just not in a convenient way. And it needs tests, actual end to end tests that look at pixels being produced with and without the user override being supplied.

**Comment by alexgleith** [comment]
Just checking a little further, and the Red band (`s3://usgs-landsat/collection02/level-2/standard/oli-tirs/2025/126/059/LC09_L2SP_126059_20250102_20250103_02_T1/LC09_L2SP_126059_20250102_20250103_02_T1_SR_B4.TIF`) has a value set (`0.0`).

I guess it's part of the whole bitflag thing. Maybe it's possible for the first bit to be set and also other bits too. (In practice, I don't think it is.)

I can report it, but still, would be nice to have a way to workaround it.

And it doesn't explain the changed behavior between old and new rasterio/gdal.

**Comment by alexgleith** [comment]
Writing and running a test, and I got this log from Rasterio:

```
WARNING  rasterio._env:__init__.py:377 CPLE_NotSupported in 5fbdbb10-a651-4663-b38d-d32c3bcb345c.tif: PREDICTOR option is ignored for COMPRESS=None. Only valid for DEFLATE, LZW, LZMA or ZSTD
WARNING  rasterio._err:warp.py:385 CPLE_AppDefined:Value 1 in the source dataset has been changed to 0 in the destination dataset to avoid being treated as NoData. To avoid this, select a different NoData value for the destination dataset.
WARNING  rasterio._env:__init__.py:377 CPLE_NotSupported in bfcd4715-020a-4f36-84ab-39b6068fd225.tif: PREDICTOR option is ignored for COMPRESS=None. Only valid for DEFLATE, LZW, LZMA or ZSTD
WARNING  rasterio._err:warp.py:385 CPLE_AppDefined:Value 1 in the source dataset has been changed to 0 in the destination dataset to avoid being treated as NoData. To avoid this, select a different NoData value for the destination dataset.
```

**Comment by Kirill888** [comment]
@alexgleith this is a problem to be solved inside `odc-stac`, if at all that is. Happy to keep an extra test for `src_nodata_fallback is not None` case, but not the `RasterBandMetadata -> RasterLoadParams` conversion populating `src_nodata_fallback`. This should come explicitly from the loading code `odc-stac` or `datacube`. This library should not make that choice for them. That's why we have the distinction in a first place `How data is` vs `How data should be loaded`.

If anything I would go the other way and introduce an extra option that would check for discrepancies between supplied raster metadata and actual metadata present in the files during loading to either `ignore quietly`(current state), `ignore with a warning` or `fail with error`, but certainly not `try to fix quietly`, which is what you are essentially proposing.

@SpacemanPaul

**Review by SpacemanPaul — APPROVED** [review]
Approved

Our behaviour should be driven by the metadata, which is ultimately in hands of the user.  The quality of the upstream data is not a hill we need to die on.


Edit: Actually didn't see Kirill's last comment.  He is right of course.  This is the right fix, but it should be applied in odc-stac, not here in the low level library.

**Comment by alexgleith** [comment]
> If anything I would go the other way and introduce an extra option that would check for discrepancies between supplied raster metadata and actual metadata present

>should be applied in odc-stac, not here in the low level library.

I feel like if we're passing in the nodata value from metadata, it should be safe assuming that it's a good value for the fallback nodata value.

I can see that we could set the value for `src_nodata_override` in `RasterLoadParams` in the odc-stac side, but again, it feels redundant, and for the rare case that a file has no nodata set in the actual file, do we care if we're already explicitly telling it what nodata value to use?

**Comment by Kirill888** [comment]
This is not resilience it's fragility. Example for you @SpacemanPaul:

your data production pipeline is affected by some bug in GDAL that results in missing nodata values in COGs it generates. You have a data QA test that loads data across multiple scenes using `datacube.load`. That test doesn't show any issues, since metadata recorded in datacube is still valid and is now used to automatically (AND SILENTLY) "fix" those data issues that you are not aware of. And no, you can't turn it off either. You see no problems until months later someone works with downloaded data in QGIS and notices absent `nodata` markers.

Example for your @alexgleith:

Elemenmt84 or whoever is doing it now see this issue and decide the correct fix is to remove `nodata` from STAC metadata, as this is a bitmask asset and they deem that concept of `nodata` doesn't apply, i.e. you really need to look at a specific bit to be 100% sure it's not valid no single value will do. So they remove it from the metadata to match data on disk configuration. You now need to patch STAC metadata before feeding it into `odc-stac`, instead of relying on a well understood and well documented approach for MANUALLY configuring workaround for certain bands that have metadata/data issues. (*"well understood and well documented approach for workarounds"* needs to be created in `odc-stac`, not here)

Also this is exactly why user defined fusers are a thing.

**Comment by Kirill888** [comment]
> I feel like if we're passing in the nodata value from metadata, it should be safe assuming that it's a good value for the fallback nodata value.

Why do you trust that more than the data on disk? If `nodata` is set on disk and is different from what STAC supplied we WILL use value off the disk currently, your proposed change doesn't change that. So you propose an inconsistent solution that is nothing but confusing if you give it more thought: trust disk more than metadata unless this specific case when disk has no `nodata` at all... NO

**Review by Kirill888 — CHANGES_REQUESTED** [review]
`.src_nodata_fallback` should not be set from metadata, see discussion.

Extra test for externally supplied `.src_nodata_fallback` is welcome though, so long as it's not a duplicate and doesn't generate warnings.

**Comment by Kirill888** [comment]
Maybe @gadomski can provide data producer perspective here.

**Comment by alexgleith** [comment]
Closing, will explore other options in `odc-stac` client side.
